text/0000-cargo-artifact-specific-dependencies.md

```diff
@@ -0,0 +1,121 @@
+
+To solve this, we can add dependencies for a specific artifact, like so:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
+
+[lib]
+name = "myproject"
+path = "src/lib.rs"
+
+[[bin]]
+name = "myproject_cli"
+path = "src/main.rs"
+dependencies = { clap = "*" }
+```
+
+Now, `clap` is required for the binary `myproject_cli`, but not for the library `myproject`, and can be used in the same way as crate-wide dependencies. In this example, `clap` could be used from the binary artifact, but attempting to use it from the library artifact would result in an unresolved import error. This has the potential to improve compile time and reduce binary size by omitting dependencies where they're not actually required. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature would add a new `dependencies` key to the `bin`, `lib`, `test`, `bench`, and `example` sections of the `Cargo.toml` file, accepting a table describing dependencies that are used only for that artifact, in addition to the ones defined in the top-level `dependencies` and `dev-dependencies` sections. These dependencies would be specified in the same way they would be in the top-level `dependencies` section, allowing version patterns, features, and alternative sources to be specified [as described in the Cargo manual](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).
+
+Additionally, artifact-specific dependencies can be specified as a non-inline table immediately following the artifact section. This is semantically identical to declaring the dependencies inline, and the syntax would look similar to crate-wide dependencies:
+
+```toml
+[[bin]]
+name = "myproject_cli"
+
+[bin.dependencies]
+clap = "*"
+
+[lib]
+name = "myproject"
+
+[lib.dependencies] # only one library can be specified, so lib name isn't required
+# ...
+```
+
+When compiling the given artifacts, the additional dependencies would be made available and linked in the same fashion as regular dependencies, allowing them to be imported with `use` and `extern crate` statements. Attempting to use the dependency from other artifacts would fail to resolve the import, raising [E0432](https://doc.rust-lang.org/error-index.html#E0432). A new help message could be added to cargo to make it easier for the user to debug the issue - e.g. `dependency specified for artifact A is not available when compiling artifact B, did you mean to specify it as a crate dependency?`.
+
+When dependencies with the same name are specified as both an artifact-specific dependency and a crate-wide dependency, a new error will be raised by cargo, even if the dependencies are identical, indicating that this isn't allowed. This could be changed in the future, with some ideas described below in the [future possibilities](#future-possibilities) section.
```

Comment:
This could use some clarification.  Presumably by "same name" this means what we call "name in toml" or the "left side name"?   I assume Cargo's duplicate package check would still be in effect.  So this would be an error:

```toml
[dependencies]
bf = {version="1.0", package="bitflags"}

[[bin]]
# …
dependencies.bitflags = "1.0"  # ERROR: depends on crate `bitflags v1.2.1` multiple times with different names
```

but this would be allowed:

```toml
[dependencies]
bf-old = {version="0.9", package="bitflags"}

[[bin]]
# …
dependencies.bitflags = "1.0"
```

Reply:
Yes, the intention was to avoid having conflicting dependency names, while keeping the existing checks for using the same version of the same dependency twice - I've added a clarification on these cases.

As noted in the future possibilities section, it could be useful to allow duplicate dependencies with the same name in specific cases (e.g. narrowing the version requirements or enabling a specific feature) like so:

```toml
[dependencies]
regex = { version = "1.3", default_features = false, features = [ "std", "perf" ] }

[[bin]]
# ...
dependencies.regex = "1.3" # enabling all default features of the regex crate
```

For now, I intentionally left this out in favor of simplicity.